From 80143a6613e720470f348f6cb7aa83005f0ba89e Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Tue, 12 Aug 2014 22:16:53 +0200
Subject: [PATCH 1024/1024] ASoC: imx-hdmi-dma: Optimize SNDRV_PCM_TRIGGER_xxx
 handling.

Do not reset frame_idx on SNDRV_PCM_TRIGGER_xxx events. This speeds
up recovering from underrun conditions in non-PCM mode. Also don't
do a full dma re-initialisation on SNDRV_PCM_TRIGGER_PAUSE_RELEASE.
---
 sound/soc/fsl/imx-hdmi-dma.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/sound/soc/fsl/imx-hdmi-dma.c b/sound/soc/fsl/imx-hdmi-dma.c
index a6ebb16..3d41b60 100644
--- a/sound/soc/fsl/imx-hdmi-dma.c
+++ b/sound/soc/fsl/imx-hdmi-dma.c
@@ -883,6 +883,7 @@ static int hdmi_dma_hw_params(struct snd_pcm_substream *substream,
 	init_table(priv->channels);
 
 	priv->appl_bytes = 0;
+	priv->frame_idx = 0;
 
 	return 0;
 }
@@ -911,8 +912,6 @@ static void hdmi_dma_trigger_init(struct snd_pcm_substream *substream,
 		iec_header.B.org_sample_freq = 0x00;
 	}
 
-	priv->frame_idx = 0;
-
 	/* Copy data by buffer_bytes */
 	hdmi_dma_data_copy(substream, priv, 'b');
 
@@ -956,13 +955,13 @@ static int hdmi_dma_trigger(struct snd_pcm_substream *substream, int cmd)
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_RESUME:
-	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		if (!check_hdmi_state())
 			return 0;
 		hdmi_dma_trigger_init(substream, priv);
 
 		dumpregs(dev);
 
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		priv->tx_active = true;
 		hdmi_audio_writeb(AHB_DMA_START, START, 0x1);
 		hdmi_dma_irq_set(false);
-- 
1.7.9.5

